pkg/kubelet/podcertificate/podcertificate_manager.go

```diff
@@ -0,0 +1,561 @@
+				continue
+			}
+
+			uniquePath := source.PodCertificate.CredentialBundlePath
+			if uniquePath == "" {
+				uniquePath = source.PodCertificate.KeyPath
+			}
+
+			key := credentialKey{
+				podUID:     pod.ObjectMeta.UID,
+				volumeName: v.Name,
+				path:       uniquePath,
+			}
+			delete(m.validCredentials, key)
+		}
+	}
+
+	klog.InfoS("XXXX ForgetPod successful completion", "pod", pod.ObjectMeta.Namespace+"/"+pod.ObjectMeta.Name)
+}
+
+// On update, check if the PodCertificateRequest has moved to a terminal state.
+// If so, convert its pending credential record to a live credential record.
+func (m *IssuingManager) handlePCRUpdate(old, new any) {
+	key, err := cache.MetaNamespaceKeyFunc(new)
+	if err != nil {
+		return
+	}
+	newPCR := new.(*certificatesv1alpha1.PodCertificateRequest)
+
+	klog.InfoS("Handling PodCertificateRequest update", "pcr", key)
+
+	state := ""
+	reason := ""
+	message := ""
+	for _, cond := range newPCR.Status.Conditions {
+		// PCRs can have only one condition
+		state = cond.Type
+		reason = cond.Reason
+		message = cond.Message
+	}
+	if state == "" {
+		// PCR has not moved to a terminal state.  Do nothing.
+		return
+	}
+
+	klog.InfoS("Checked state of PodCertificateRequest", "state", state)
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	// Clear the pending credential record and make a live credential
+	// record.
+	pendingRecord := m.pendingCredentials[key]
+	delete(m.pendingCredentials, key)
+
+	if pendingRecord == nil {
+		// If we don't have a pending record corresponding to this PCR, there
+		// are two possible causes:
+		//
+		// 1) Some admin manually created a PCR that targets a pod on our node.
+		//
+		// 2) Kubelet created this PCR, but then restarted.
+		//
+		// In both cases, just do nothing.
+		klog.InfoS("PodCertificateRequest update was for stray", "pcr", key)
+		return
+	}
+
+	finalKey := credentialKey{
+		podUID:     newPCR.Spec.PodUID,
+		volumeName: pendingRecord.volumeName,
+		path:       pendingRecord.path,
+	}
+	if state == certificatesv1alpha1.PodCertificateRequestConditionTypeIssued {
+		klog.InfoS("Recording valid credential entry", "poduid", newPCR.Spec.PodUID, "volumename", pendingRecord.volumeName, "path", pendingRecord.path)
+		m.validCredentials[finalKey] = &validCredentialRecord{
+			key:            pendingRecord.privateKey,
+			certificates:   []byte(newPCR.Status.CertificateChain),
+			beginRefreshAt: newPCR.Status.BeginRefreshAt.Time,
+			expiresAt:      newPCR.Status.NotAfter.Time,
+		}
+	} else {
+		klog.InfoS("Recording invalid credential entry", "poduid", newPCR.Spec.PodUID, "volumename", pendingRecord.volumeName, "path", pendingRecord.path)
+		m.invalidCredentials[finalKey] = &invalidCredentialRecord{
+			condition: state,
+			reason:    reason,
+			message:   message,
+		}
+	}
+}
+
+// On delete, clean up the pending credential record corresponding
+// to this PodCertificateRequest, if any exists.
+func (m *IssuingManager) handlePCRDelete(old any) {
```

Comment:
If I understand it correctly, `kubectl delete podcertificaterequest -A --all` will make all pending pods requesting a cert never to come up. Is that ok?